MINIFICPP-1062 - Improve FlowFile repository performance by leveraging batching

arpadboda opened the pull request:
Thank you for submitting a contribution to Apache NiFi - MiNiFi C++.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced
     in the commit message?

- [ ] Does your PR title start with MINIFICPP-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ ] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [ ] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE file?
- [ ] If applicable, have you updated the NOTICE file?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

arpadboda commented:
@achristianson you may want to a look at this

achristianson commented:
I think it may be unrelated, but I am getting this error which I haven't seen before:

```
make[2]: *** No rule to make target 'thirdparty/curl-install/lib64/libcurl.a', needed by 'main/minifi'.  Stop.
make[2]: *** Waiting for unfinished jobs....
```

arpadboda commented:
> I think it may be unrelated, but I am getting this error which I haven't seen before:
> 
> ```
> make[2]: *** No rule to make target 'thirdparty/curl-install/lib64/libcurl.a', needed by 'main/minifi'.  Stop.
> make[2]: *** Waiting for unfinished jobs....
> ```

I think you should try clean build.

Unrelated, but if you have performance issues, it's quite an easy win, I think it adds way more than nmap.

achristianson commented:
This is on a clean build

arpadboda commented:
> This is on a clean build

Did you bootstrap? This change has nothing to do with curl.

phrocker commented on extensions/rocksdb-repos/FlowFileRepository.h:
```diff
@@ -122,6 +126,21 @@ class FlowFileRepository : public core::Repository, public std::enable_shared_fr
       return false;
   }
+
+  virtual bool MultiPut(const std::vector<std::tuple<std::string, const uint8_t *, size_t>> data) {
```
this looks like good functionality. Why is this WIP? have you proven it in terms of performance improvements?

arpadboda replied:
Yes, it's way faster, will share some results. 

WIP as I'm not yet happy with the API, plan some improvements before considered as final.

bakaid commented on extensions/rocksdb-repos/FlowFileRepository.cpp:
```diff
@@ -32,25 +33,34 @@ namespace repository {
+
+  for(size_t i=0; i<keys.size() && i<values.size(); ++i) {
+    decrement_total += values[i].size();
+    std::shared_ptr<FlowFileRecord> eventRead = std::make_shared<FlowFileRecord>(shared_from_this(), content_repo_);
```
MultiGet returns a vector of statuses, indicating for each individual key whether it was found. If it was not, the corresponding value `will be set to some arbitrary value (often "")` (https://github.com/facebook/rocksdb/blob/master/include/rocksdb/db.h#L430).
While I assume that not finding a key in this case would indicate some kind of inconsistency in the database caused by us, I am unfortunately not at all sure that this can not happen. In this case we would be trying to `DeSerialize` an arbitrary string as an event, which is not too good.
I would prefer this to be rewritten to take into account `MultiGet`'s return value and log an error if a key was not found, but do not try to do anything with the corresponding value.

arpadboda replied:
Thanks, fair point, added the check!

bakaid commented on extensions/rocksdb-repos/FlowFileRepository.cpp:
```diff
@@ -32,25 +33,34 @@ namespace repository {
-      }
-      logger_->log_debug("Issuing batch delete, including %s, Content path %s", eventRead->getUUIDStr(), eventRead->getContentFullPath());
-      batch.Delete(key);
+      keys.push_back(key);
```
I think this can be moved here.

arpadboda replied:
Done

bakaid commented on libminifi/test/unit/ProvenanceTestHelper.h:
```diff
@@ -71,11 +71,24 @@ class TestRepository : public core::Repository {
     return true;
   }
 
+  bool MultiPut(const std::vector<std::tuple<std::string, const uint8_t *, size_t>> data) {
```
const& please.

arpadboda replied:
Done

bakaid commented:
> Compiled the code and ran the tests, which worked fine, but running an actual GenerateFlowFile (batch of 10) -> LogAttribute flow, while works, generates the following logs:
> 
> ```
>                                                                                                                                       ��! DB is most probably in an inconsistent state!
> [2019-11-11 17:32:08.869] [org::apache::nifi::minifi::core::repository::FlowFileRepository] [error] Failed to read key from rocksdb: ���! DB is most probably in an inconsistent state!
> [2019-11-11 17:32:08.869] [org::apache::nifi::minifi::core::repository::FlowFileRepository] [error] Failed to read key from rocksdb: o��! DB is most probably in an inconsistent state!
> [2019-11-11 17:32:08.869] [org::apache::nifi::minifi::core::repository::FlowFileRepository] [error] Failed to read key from rocksdb: ! DB is most probably in an inconsistent state!
> [2019-11-11 17:32:08.869] [org::apache::nifi::minifi::core::repository::FlowFileRepository] [error] Failed to read key from rocksdb: ! DB is most probably in an inconsistent state!
> [2019-11-11 17:32:08.869] [org::apache::nifi::minifi::core::repository::FlowFileRepository] [error] Failed to read key from rocksdb: ! DB is most probably in an inconsistent state!
> [2019-11-11 17:32:08.869] [org::apache::nifi::minifi::core::repository::FlowFileRepository] [error] Failed to read key from rocksdb: ! DB is most probably in an inconsistent state!
> ```
> 
> Which is disconcerting both because of hitting this condition and the memory garbage as keys.
> There might be some eventual consistency or transaction magic which can account for the first, but then we have to revise the logging.
> 
> This has to be investigated.

Platform: macOS
config.yml:
```MiNiFi Config Version: 3
Flow Controller:
  name: elclass7
  comment: Created by MiNiFi C2 Flow Designer
Core Properties:
  flow controller graceful shutdown period: 10 sec
  flow service write delay interval: 500 ms
  administrative yield duration: 30 sec
  bored yield duration: 10 millis
  max concurrent threads: 1
  variable registry properties: ''
FlowFile Repository:
  partitions: 256
  checkpoint interval: 2 mins
  always sync: false
  Swap:
    threshold: 20000
    in period: 5 sec
    in threads: 1
    out period: 5 sec
    out threads: 4
Content Repository:
  content claim max appendable size: 10 MB
  content claim max flow files: 100
  always sync: false
Provenance Repository:
  provenance rollover time: 1 min
  implementation: org.apache.nifi.provenance.MiNiFiPersistentProvenanceRepository
Component Status Repository:
  buffer size: 1440
  snapshot frequency: 1 min
Security Properties:
  keystore: ''
  keystore type: ''
  keystore password: ''
  key password: ''
  truststore: ''
  truststore type: ''
  truststore password: ''
  ssl protocol: ''
  Sensitive Props:
    key:
    algorithm: PBEWITHMD5AND256BITAES-CBC-OPENSSL
    provider: BC
Processors:
- id: a0a92728-54aa-4f22-875e-287244a4422c
  name: GenerateFlowFile
  class: org.apache.nifi.minifi.processors.GenerateFlowFile
  max concurrent tasks: 1
  scheduling strategy: TIMER_DRIVEN
  scheduling period: 1000 ms
  penalization period: 30000 ms
  yield period: 1000 ms
  run duration nanos: 0
  auto-terminated relationships list: []
  Properties:
    Batch Size: '1'
    Data Format: Text
    File Size: 1 kB
    Unique FlowFiles: 'true'
- id: 200b4e82-0edd-4cfa-b8c2-f9f86fa027a6
  name: PutFile
  class: org.apache.nifi.minifi.processors.PutFile
  max concurrent tasks: 1
  scheduling strategy: TIMER_DRIVEN
  scheduling period: 1000 ms
  penalization period: 30000 ms
  yield period: 1000 ms
  run duration nanos: 0
  auto-terminated relationships list:
  - failure
  - success
  Properties:
    Conflict Resolution Strategy: fail
    Create Missing Directories: 'true'
    Directory: /tmp/put/${now():format('Y')}
    Maximum File Count: '-1'
Controller Services: []
Process Groups: []
Input Ports: []
Output Ports: []
Funnels: []
Connections:
- id: d0fc88e8-f0e4-44fc-b20b-225a8ce7a787
  name: GenerateFlowFile/success/PutFile
  source id: a0a92728-54aa-4f22-875e-287244a4422c
  source relationship names:
  - success
  destination id: 200b4e82-0edd-4cfa-b8c2-f9f86fa027a6
  max work queue size: 0
  max work queue data size: 10000 B
  flowfile expiration: 0 seconds
  queue prioritizer class: ''
Remote Process Groups: []
NiFi Properties Overrides: {}```

bakaid commented:
> Compiled the code and ran the tests, which worked fine, but running an actual GenerateFlowFile (batch of 10) -> LogAttribute flow, while works, generates the following logs:
> 
> ```
>                                                                                                                                       ��! DB is most probably in an inconsistent state!
> [2019-11-11 17:32:08.869] [org::apache::nifi::minifi::core::repository::FlowFileRepository] [error] Failed to read key from rocksdb: ���! DB is most probably in an inconsistent state!
> [2019-11-11 17:32:08.869] [org::apache::nifi::minifi::core::repository::FlowFileRepository] [error] Failed to read key from rocksdb: o��! DB is most probably in an inconsistent state!
> [2019-11-11 17:32:08.869] [org::apache::nifi::minifi::core::repository::FlowFileRepository] [error] Failed to read key from rocksdb: ! DB is most probably in an inconsistent state!
> [2019-11-11 17:32:08.869] [org::apache::nifi::minifi::core::repository::FlowFileRepository] [error] Failed to read key from rocksdb: ! DB is most probably in an inconsistent state!
> [2019-11-11 17:32:08.869] [org::apache::nifi::minifi::core::repository::FlowFileRepository] [error] Failed to read key from rocksdb: ! DB is most probably in an inconsistent state!
> [2019-11-11 17:32:08.869] [org::apache::nifi::minifi::core::repository::FlowFileRepository] [error] Failed to read key from rocksdb: ! DB is most probably in an inconsistent state!
> ```
> 
> Which is disconcerting both because of hitting this condition and the memory garbage as keys.
> There might be some eventual consistency or transaction magic which can account for the first, but then we have to revise the logging.
> 
> This has to be investigated.

@arpadboda 
Platform: macOS
config.yml:
```danielbakai@BakaiDL-MBP15 ~/test2 $ cat nifi-minifi-cpp-0.7.0/conf/config.yml
MiNiFi Config Version: 3
Flow Controller:
  name: aclass
  comment: Created by MiNiFi C2 Flow Designer
Core Properties:
  flow controller graceful shutdown period: 10 sec
  flow service write delay interval: 500 ms
  administrative yield duration: 30 sec
  bored yield duration: 10 millis
  max concurrent threads: 1
  variable registry properties: ''
FlowFile Repository:
  partitions: 256
  checkpoint interval: 2 mins
  always sync: false
  Swap:
    threshold: 20000
    in period: 5 sec
    in threads: 1
    out period: 5 sec
    out threads: 4
Content Repository:
  content claim max appendable size: 10 MB
  content claim max flow files: 100
  always sync: false
Provenance Repository:
  provenance rollover time: 1 min
  implementation: org.apache.nifi.provenance.MiNiFiPersistentProvenanceRepository
Component Status Repository:
  buffer size: 1440
  snapshot frequency: 1 min
Security Properties:
  keystore: ''
  keystore type: ''
  keystore password: ''
  key password: ''
  truststore: ''
  truststore type: ''
  truststore password: ''
  ssl protocol: ''
  Sensitive Props:
    key:
    algorithm: PBEWITHMD5AND256BITAES-CBC-OPENSSL
    provider: BC
Processors:
- id: 2f77d1c7-4a6a-4dab-b16a-5118eef13c4d
  name: GenerateFlowFile
  class: org.apache.nifi.minifi.processors.GenerateFlowFile
  max concurrent tasks: 1
  scheduling strategy: TIMER_DRIVEN
  scheduling period: 1000 ms
  penalization period: 30000 ms
  yield period: 1000 ms
  run duration nanos: 0
  auto-terminated relationships list: []
  Properties:
    Batch Size: '2'
    Data Format: Text
    File Size: 1 kB
    Unique FlowFiles: 'true'
- id: 88b3f926-74e1-4047-b297-5dee4eaa9a12
  name: LogAttribute
  class: org.apache.nifi.minifi.processors.LogAttribute
  max concurrent tasks: 1
  scheduling strategy: EVENT_DRIVEN
  scheduling period: 1000 ms
  penalization period: 30000 ms
  yield period: 1000 ms
  run duration nanos: 0
  auto-terminated relationships list:
  - success
  Properties:
    FlowFiles To Log: '0'
    Hexencode Payload: 'false'
    Log Payload: 'true'
    Maximum Payload Line Length: '80'
Controller Services: []
Process Groups: []
Input Ports: []
Output Ports: []
Funnels: []
Connections:
- id: 0108c38a-417b-40dc-83c8-d35c4d790662
  name: GenerateFlowFile/success/LogAttribute
  source id: 2f77d1c7-4a6a-4dab-b16a-5118eef13c4d
  source relationship names:
  - success
  destination id: 88b3f926-74e1-4047-b297-5dee4eaa9a12
  max work queue size: 0
  max work queue data size: 10000 B
  flowfile expiration: 0 seconds
  queue prioritizer class: ''
Remote Process Groups: []
NiFi Properties Overrides: {}
```

arpadboda commented:
> > Compiled the code and ran the tests, which worked fine, but running an actual GenerateFlowFile (batch of 10) -> LogAttribute flow, while works, generates the following logs:
> > ```
> >                                                                                                                                       ��! DB is most probably in an inconsistent state!
> > [2019-11-11 17:32:08.869] [org::apache::nifi::minifi::core::repository::FlowFileRepository] [error] Failed to read key from rocksdb: ���! DB is most probably in an inconsistent state!
> > [2019-11-11 17:32:08.869] [org::apache::nifi::minifi::core::repository::FlowFileRepository] [error] Failed to read key from rocksdb: o��! DB is most probably in an inconsistent state!
> > [2019-11-11 17:32:08.869] [org::apache::nifi::minifi::core::repository::FlowFileRepository] [error] Failed to read key from rocksdb: ! DB is most probably in an inconsistent state!
> > [2019-11-11 17:32:08.869] [org::apache::nifi::minifi::core::repository::FlowFileRepository] [error] Failed to read key from rocksdb: ! DB is most probably in an inconsistent state!
> > [2019-11-11 17:32:08.869] [org::apache::nifi::minifi::core::repository::FlowFileRepository] [error] Failed to read key from rocksdb: ! DB is most probably in an inconsistent state!
> > [2019-11-11 17:32:08.869] [org::apache::nifi::minifi::core::repository::FlowFileRepository] [error] Failed to read key from rocksdb: ! DB is most probably in an inconsistent state!
> > ```
> > 
> > 
> > Which is disconcerting both because of hitting this condition and the memory garbage as keys.
> > There might be some eventual consistency or transaction magic which can account for the first, but then we have to revise the logging.
> > This has to be investigated.
> 
> @arpadboda
> Platform: macOS
> config.yml:
> 
> ```
> MiNiFi Config Version: 3
> Flow Controller:
>   name: aclass
>   comment: Created by MiNiFi C2 Flow Designer
> Core Properties:
>   flow controller graceful shutdown period: 10 sec
>   flow service write delay interval: 500 ms
>   administrative yield duration: 30 sec
>   bored yield duration: 10 millis
>   max concurrent threads: 1
>   variable registry properties: ''
> FlowFile Repository:
>   partitions: 256
>   checkpoint interval: 2 mins
>   always sync: false
>   Swap:
>     threshold: 20000
>     in period: 5 sec
>     in threads: 1
>     out period: 5 sec
>     out threads: 4
> Content Repository:
>   content claim max appendable size: 10 MB
>   content claim max flow files: 100
>   always sync: false
> Provenance Repository:
>   provenance rollover time: 1 min
>   implementation: org.apache.nifi.provenance.MiNiFiPersistentProvenanceRepository
> Component Status Repository:
>   buffer size: 1440
>   snapshot frequency: 1 min
> Security Properties:
>   keystore: ''
>   keystore type: ''
>   keystore password: ''
>   key password: ''
>   truststore: ''
>   truststore type: ''
>   truststore password: ''
>   ssl protocol: ''
>   Sensitive Props:
>     key:
>     algorithm: PBEWITHMD5AND256BITAES-CBC-OPENSSL
>     provider: BC
> Processors:
> - id: 2f77d1c7-4a6a-4dab-b16a-5118eef13c4d
>   name: GenerateFlowFile
>   class: org.apache.nifi.minifi.processors.GenerateFlowFile
>   max concurrent tasks: 1
>   scheduling strategy: TIMER_DRIVEN
>   scheduling period: 1000 ms
>   penalization period: 30000 ms
>   yield period: 1000 ms
>   run duration nanos: 0
>   auto-terminated relationships list: []
>   Properties:
>     Batch Size: '10'
>     Data Format: Text
>     File Size: 1 kB
>     Unique FlowFiles: 'true'
> - id: 88b3f926-74e1-4047-b297-5dee4eaa9a12
>   name: LogAttribute
>   class: org.apache.nifi.minifi.processors.LogAttribute
>   max concurrent tasks: 1
>   scheduling strategy: EVENT_DRIVEN
>   scheduling period: 1000 ms
>   penalization period: 30000 ms
>   yield period: 1000 ms
>   run duration nanos: 0
>   auto-terminated relationships list:
>   - success
>   Properties:
>     FlowFiles To Log: '0'
>     Hexencode Payload: 'false'
>     Log Payload: 'true'
>     Maximum Payload Line Length: '80'
> Controller Services: []
> Process Groups: []
> Input Ports: []
> Output Ports: []
> Funnels: []
> Connections:
> - id: 0108c38a-417b-40dc-83c8-d35c4d790662
>   name: GenerateFlowFile/success/LogAttribute
>   source id: 2f77d1c7-4a6a-4dab-b16a-5118eef13c4d
>   source relationship names:
>   - success
>   destination id: 88b3f926-74e1-4047-b297-5dee4eaa9a12
>   max work queue size: 0
>   max work queue data size: 10000 B
>   flowfile expiration: 0 seconds
>   queue prioritizer class: ''
> Remote Process Groups: []
> NiFi Properties Overrides: {}
> ```

@bakaid : 
Thanks, very good spot, fixed.

So the problem was rocksdb::Slice ctor:
```
  Slice(const std::string& s) : data_(s.data()), size_(s.size()) { }
```

Which doesn't copy the content of the string, just grabs ptrs. 
This is one of the most evil APIs I have ever seen. It's simply a landmine. 

This was wrong by the time I made this change, but the comment to move the string ( https://github.com/apache/nifi-minifi-cpp/pull/664#discussion_r344624115 ) made it even worse as it further reduced the lifetime of the string. 

Now I fixed it, run MiNiFi for a while and didn't see this error logged any more. (Before I could reproduce easily as well)
